bin/propolis-server/src/lib/spec/api_spec_v0.rs

```diff
@@ -0,0 +1,259 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Conversions from version-0 instance specs in the [`propolis_api_types`]
+//! crate to the internal [`super::Spec`] representation.
+
+use propolis_api_types::instance_spec::{
+    components::devices::{SerialPort as SerialPortDesc, SerialPortNumber},
+    v0::{InstanceSpecV0, NetworkBackendV0, NetworkDeviceV0, StorageDeviceV0},
+};
+use thiserror::Error;
+
+#[cfg(feature = "falcon")]
+use propolis_api_types::instance_spec::components::devices::SoftNpuPort as SoftNpuPortSpec;
+
+#[cfg(feature = "falcon")]
+use crate::spec::SoftNpuPort;
+
+use super::{
+    builder::{SpecBuilder, SpecBuilderError},
+    Disk, Nic, QemuPvpanic, SerialPortUser, Spec,
+};
+
+#[derive(Debug, Error)]
+pub(crate) enum ApiSpecParseError {
+    #[error(transparent)]
+    BuilderError(#[from] SpecBuilderError),
+
+    #[error("backend {backend} not found for device {device}")]
+    BackendNotFound { backend: String, device: String },
+
+    #[error("backend {0} not used by any device")]
+    BackendNotUsed(String),
+
+    #[error("network backend for guest NIC {0} is not a viona backend")]
+    GuestNicInvalidBackend(String),
+
+    #[cfg(feature = "falcon")]
+    #[error("network backend for device {0} is not a DLPI backend")]
+    NotDlpiBackend(String),
+}
+
+impl From<Spec> for InstanceSpecV0 {
+    fn from(val: Spec) -> Self {
+        let mut spec = InstanceSpecV0::default();
+
+        spec.devices.board = val.board;
+        for (disk_name, disk) in val.disks {
```

Comment:
it occurs to me just now that we might want to stick something like this here:
```suggestion
        spec.devices.storage_devices.reserve(val.disks.len());
        spec.devices.storage_backends.reserve(val.disks.len());
        for (disk_name, disk) in val.disks {
```
Since we know the number of devices up front, we may as well preallocate enough capacity for them, rather than repeatedly growing the collections as we insert more devices. On the other hand, this is probably not a huge deal --- the improved performance is probably marginal at best, and I'm sure constructing these specs isn't what dominates instance-creation time anyway.

Reply:
(if you do decide to do this, we could also do it for the other loops as well)

Reply:
I've also left this one alone for now since this type is going to get a substantial makeover very soon.